### ISSUE TITLE
`different number of opening and closing braces` when opening GTDB tree

### ISSUE THREAD

**Issue by MrCreosote** [glottobank/python-newick#50]
Hi, I'm trying to open the most recently published [GTDB](https://gtdb.ecogenomic.org/) [bacterial](https://data.gtdb.ecogenomic.org/releases/release207/207.0/bac120_r207.tree.tar.gz) and [archeal](https://data.gtdb.ecogenomic.org/releases/release207/207.0/ar53_r207.tree.tar.gz) trees, and `python-newick` is failing, noting that the counts of opening and closing parentheses are unequal. However, counting them manually looks ok:

```python
In [1]: import newick

In [2]: newick.__version__
Out[2]: '1.4.0'

In [3]: newick.read('ar53_r207.tree')
*snip*

File ~/.local/share/virtualenvs/py3.9playground--W4ZB8cD/lib/python3.9/site-packages/newick.py:511, in parse_node(s, strip_comments, **kw)
    509 parts = s.split(')')
    510 if len(parts) != len(s.split('(')):
--> 511     raise ValueError('different number of opening and closing braces')
    512 if len(parts) == 1:
    513     descendants, label = [], s

ValueError: different number of opening and closing braces

In [4]: newick.read('bac120_r207.tree')
*snip*

File ~/.local/share/virtualenvs/py3.9playground--W4ZB8cD/lib/python3.9/site-packages/newick.py:511, in parse_node(s, strip_comments, **kw)
    509 parts = s.split(')')
    510 if len(parts) != len(s.split('(')):
--> 511     raise ValueError('different number of opening and closing braces')
    512 if len(parts) == 1:
    513     descendants, label = [], s

ValueError: different number of opening and closing braces

In [5]: with open('ar53_r207.tree') as t:
   ...:     tree = t.read()
   ...: 

In [6]: tree.count('(')
Out[6]: 3411

In [7]: tree.count(')')
Out[7]: 3411

In [8]: with open('bac120_r207.tree') as t:
   ...:     tree = t.read()
   ...: 

In [9]: tree.count('(')
Out[9]: 62290

In [10]: tree.count(')')
Out[10]: 62290
```

**Comment by xrotwang** [comment]
My first guess would be the colons in the taxa labels. Newick is
unfortunately a bit underspecified when it comes to allowed characters in
labels, etc.
Will have to investigate when I'm back at my computer.

MrCreosote ***@***.***> schrieb am So., 25. Dez. 2022, 20:35:

**Comment by MrCreosote** [comment]
E.g. `'100.0:g__Halorubrum':0.06022`? I'm guessing those are supposed to be the bootstrap support values...? I wouldn't expect them to be inside the single quotes though

**Comment by xrotwang** [comment]
Ok, I think I found the issue. The tree has `;` in ~150 quoted node labels, and `python-newick` interprets `;` unconditionally as separator of subtrees. I think that behaviour is "correct" in the sense that it doesn't make full sense to look for quotes when we don't know what's a tree or what's a node yet. It might still be possible (or even simple) to implement parsing that only uses semicolons outside of quotes. Will look into this.

**Comment by MrCreosote** [comment]
Sounds like the colons are moot, but apparently there are implementations that do indeed store the bootstrap values as part of the node name, which is just... wow
https://forum.gtdb.ecogenomic.org/t/gtdbtk-meaning-of-numbers-in-internal-node-names-in-the-output-tree/258/2

**Comment by xrotwang** [comment]
I just got around to implement more full-fledged support for quoted labels. With the code in #51 quoted labels may contain anything (as long as single quotes are escaped properly). This slows down parsing quite a bit, but I guess it's worth it.

**Comment by xrotwang** [comment]
@MrCreosote just released [newick 1.5](https://pypi.org/project/newick/1.5.0/), which should have no problems reading the GTDB data.
